src/lib/hash/sha2_64/sha2_64_armv8/sha2_64_armv8.cpp

```diff
@@ -0,0 +1,407 @@
+      STATE2 = vsha512h2q_u64(TSTATE1, STATE0, STATE3);
+      STATE0 = vaddq_u64(STATE0, TSTATE1);
+      MSG1 = vsha512su1q_u64(vsha512su0q_u64(MSG1, MSG2), MSG0, vextq_u64(MSG5, MSG6, 1));
+
+      // Rounds 4-5
+      MSG_K = vaddq_u64(MSG2, vld1q_u64(&K[2 * 2]));
+      TSTATE0 = vaddq_u64(vextq_u64(MSG_K, MSG_K, 1), STATE1);
+      TSTATE1 = vsha512hq_u64(TSTATE0, vextq_u64(STATE0, STATE1, 1), vextq_u64(STATE3, STATE0, 1));
+      STATE1 = vsha512h2q_u64(TSTATE1, STATE3, STATE2);
+      STATE3 = vaddq_u64(STATE3, TSTATE1);
+      MSG2 = vsha512su1q_u64(vsha512su0q_u64(MSG2, MSG3), MSG1, vextq_u64(MSG6, MSG7, 1));
+
+      // Rounds 6-7
+      MSG_K = vaddq_u64(MSG3, vld1q_u64(&K[2 * 3]));
+      TSTATE0 = vaddq_u64(vextq_u64(MSG_K, MSG_K, 1), STATE0);
+      TSTATE1 = vsha512hq_u64(TSTATE0, vextq_u64(STATE3, STATE0, 1), vextq_u64(STATE2, STATE3, 1));
+      STATE0 = vsha512h2q_u64(TSTATE1, STATE2, STATE1);
+      STATE2 = vaddq_u64(STATE2, TSTATE1);
+      MSG3 = vsha512su1q_u64(vsha512su0q_u64(MSG3, MSG4), MSG2, vextq_u64(MSG7, MSG0, 1));
+
+      // Rounds 8-9
+      MSG_K = vaddq_u64(MSG4, vld1q_u64(&K[2 * 4]));
+      TSTATE0 = vaddq_u64(vextq_u64(MSG_K, MSG_K, 1), STATE3);
+      TSTATE1 = vsha512hq_u64(TSTATE0, vextq_u64(STATE2, STATE3, 1), vextq_u64(STATE1, STATE2, 1));
+      STATE3 = vsha512h2q_u64(TSTATE1, STATE1, STATE0);
+      STATE1 = vaddq_u64(STATE1, TSTATE1);
+      MSG4 = vsha512su1q_u64(vsha512su0q_u64(MSG4, MSG5), MSG3, vextq_u64(MSG0, MSG1, 1));
+
+      // Rounds 10-11
+      MSG_K = vaddq_u64(MSG5, vld1q_u64(&K[2 * 5]));
+      TSTATE0 = vaddq_u64(vextq_u64(MSG_K, MSG_K, 1), STATE2);
+      TSTATE1 = vsha512hq_u64(TSTATE0, vextq_u64(STATE1, STATE2, 1), vextq_u64(STATE0, STATE1, 1));
+      STATE2 = vsha512h2q_u64(TSTATE1, STATE0, STATE3);
+      STATE0 = vaddq_u64(STATE0, TSTATE1);
+      MSG5 = vsha512su1q_u64(vsha512su0q_u64(MSG5, MSG6), MSG4, vextq_u64(MSG1, MSG2, 1));
+
+      // Rounds 12-13
+      MSG_K = vaddq_u64(MSG6, vld1q_u64(&K[2 * 6]));
+      TSTATE0 = vaddq_u64(vextq_u64(MSG_K, MSG_K, 1), STATE1);
+      TSTATE1 = vsha512hq_u64(TSTATE0, vextq_u64(STATE0, STATE1, 1), vextq_u64(STATE3, STATE0, 1));
+      STATE1 = vsha512h2q_u64(TSTATE1, STATE3, STATE2);
+      STATE3 = vaddq_u64(STATE3, TSTATE1);
+      MSG6 = vsha512su1q_u64(vsha512su0q_u64(MSG6, MSG7), MSG5, vextq_u64(MSG2, MSG3, 1));
+
+      // Rounds 14-15
+      MSG_K = vaddq_u64(MSG7, vld1q_u64(&K[2 * 7]));
+      TSTATE0 = vaddq_u64(vextq_u64(MSG_K, MSG_K, 1), STATE0);
+      TSTATE1 = vsha512hq_u64(TSTATE0, vextq_u64(STATE3, STATE0, 1), vextq_u64(STATE2, STATE3, 1));
+      STATE0 = vsha512h2q_u64(TSTATE1, STATE2, STATE1);
+      STATE2 = vaddq_u64(STATE2, TSTATE1);
+      MSG7 = vsha512su1q_u64(vsha512su0q_u64(MSG7, MSG0), MSG6, vextq_u64(MSG3, MSG4, 1));
+
+      ////// UNROLLED LOOP //////
+
+      // TODOs:
+      // tests with long inputs fail
```

Comment:
?!?

Reply:
Embarrassing leftover from debugging, removed it.